### ISSUE TITLE
Mockito annotations support?

### ISSUE THREAD

**Issue by ishustava** [greghaskins/spectrum#34]
We have found that using `@Mock` and `@InjectMocks` annotations don't work with spectrum, resulting in the mock causing a `NullPointerException`.

It seems that the runner needs to do all the injections and mocking pretty early in the test execution, so that all instances are injected and/or mocked.

This issue is somewhat related to [this issue](https://github.com/greghaskins/spectrum/issues/15), but perhaps it's helpful to bring it up here separately.

Thanks for the great framework,

@ishustava && @larham

**Comment by greghaskins** [comment]
Thanks for filing an issue @ishustava and @larham, it is somewhat different from #15 and worth tracking. `@Mock` and `@InjectMocks` only work with `MocktioJUnitRunner`, and probably won't be able to be integrated out-of-the-box with Spectrum.

Making Spectrum easier to work with Mocktio, however, is something that we should consider. It wouldn't look exactly the same, but would provide some of the same convenience. I have been thinking that another helper/wrapper could be useful, perhaps to run in a `beforeEach` block to wire up mocks. That would probably be a related, but separate project. 

Just sketching out ideas, what do you think of something like this?

```java
describe("a cash register", () -> {

	Value<CashRegister> cashRegister = value();
	Value<Drawer> drawer = value();
	Value<BarcodeScanner> scanner = value();

	beforeEach(() -> {
		injectMocks(drawer, scanner).into(cashRegister));
	});
	
	it("should open the drawer to accept payment", () -> {
		cashRegister.acceptPayment();
		
		verify(drawer).open();
	});
	
	it("should read from the barcode scanner", () -> {
		when(scanner.read()).thenReturn(123456);
		
		cashRegister.readNextItem();
		
		expect(cashRegister.lastItem()).toBe(123456);
	});

});
```

**Comment by larham** [comment]
in our attempts so far, injection can be handled with beforeEach(() -> MockitoAnnotations.injectMocks(this));

**Comment by larham** [comment]
```@RunWith(Spectrum.class)
public class SampleMockito {
​
  @Mock
  Environment environment;
​
  @InjectMocks
  Thing subject;
​
  beforeEach(() -> MockitoAnnotations.injectMocks(this));
​
  describe("something", ...)
}
```

**Comment by larham** [comment]
is there any magic that could detect the presence of the Mockito annotations and add that "before" injection under the hood?

**Comment by greghaskins** [comment]
Thanks for sharing the example. I understand better now. 

Unfortunately, I think we'll have to find another way for mocks and injection to work more seamlessly with Spectrum. The Mockito annotations are very tightly tied to the traditional way of writing JUnit tests, using fields and methods instead of nested function closures, like RSpec, Jasmine, and Spectrum. With BDD-style runners, suites can have arbitrary levels of nesting and composition, which is part of their power.

We could write some syntactic sugar to achieve the same effect as your example, but it would break the nesting model since you couldn't declare or inject mocks from inside a nested suite. Those fields only apply at the top level.

I did some digging to see if there was a way this problem was solved already on other platforms.

- RSpec has a mocking library, but doesn't (as far as I can tell) have a magic inject-all-the-mocks feature. 
- Similarly, Jasmine spies can be created and verified similar to Mockito, but they also don't have a way to automatically instantitate using mocks.
- The AngularJS project uses a global dependency injector for *everything*, which makes it fairly easy to inject mocks for testing. This approach, however, requires a very specific, global way of resolving runtime dependencies.

In your projection code, are you using Spring or Guice (or something similar) for dependency injection? If so, there may be more options.

Otherwise, we can sketch up an API for providing an equivalent `@InjectMocks` feature in a way that works with Spectrum.
